### PR TITLE
451006 - Denis Ponyakov

### PR DESCRIPTION
**Pull request by nexon-97**
Finished lab, and would appreciate some code review.

### CODE REVIEW

File: AppSample/BasicTreeNode.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Collections.Generic;
+
+public class BasicTreeNode<T>
+{
+	public List<BasicTreeNode<T>> Children = new List<BasicTreeNode<T>>();
```
Comment (RamanBut-Husaim) [inline AppSample/BasicTreeNode.cs]: @nexon-97 public fields? Horrible. :) Use properties instead.

File: AppSample/ConsoleTraceResultFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+
+public class ConsoleTraceResultFormatter : ITraceResultFormatter
+{
+	const string PaddingSpace = "  ";
```
Comment (RamanBut-Husaim) [inline AppSample/ConsoleTraceResultFormatter.cs]: @nexon-97 Try to always specify visibility modifiers.

File: AppSample/ThreadJob.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		Thread.Sleep(500);
+		foo();
+
+		UsedTracer.StopTrace();
+
+		ThreadWatch.Stop();
+		UsedTracer.SetThreadTime(ThreadWatch.ElapsedMilliseconds);
+	}
+
+	void foo()
```
Comment (RamanBut-Husaim) [inline AppSample/ThreadJob.cs]: @nexon-97 `PascalCase` for method naming.

File: AppSample/TraceResult.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Collections.Concurrent;
+
+public class TraceResult
+{
+	public class TraceComponent
+	{
+		public string MethodName;
```
Comment (RamanBut-Husaim) [inline AppSample/TraceResult.cs]: @nexon-97 Properties.

File: AppSample/TraceResult.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Collections.Concurrent;
+
+public class TraceResult
+{
+	public class TraceComponent
```
Comment (RamanBut-Husaim) [inline AppSample/TraceResult.cs]: @nexon-97 Public nested classes? Change visibility modifier if possible or move to separate files.

File: AppSample/XmlTraceResultFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
+﻿using System.Xml;
+
+public class XmlTraceResultFormatter : ITraceResultFormatter
+{
+	public string OutputFileName = "TracerOutput.xml";
```
Comment (RamanBut-Husaim) [inline AppSample/XmlTraceResultFormatter.cs]: @nexon-97 Move to a constructor parameter and pass the constant from the outside world.

File: AppSample/TraceResult.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+		public BasicTreeNode<TraceComponent> RootComponent = null;
+		public BasicTreeNode<TraceComponent> CurrentNode = null;
+		public long ExecutionTime;
+	}
+
+	public ConcurrentDictionary<int, ThreadData> ThreadsData = new ConcurrentDictionary<int, ThreadData>();
+
+	public void StartComponent(string methodName, int paramCount, string className)
+	{
+		int ThreadId = Thread.CurrentThread.ManagedThreadId;
```
Comment (RamanBut-Husaim) [inline AppSample/TraceResult.cs]: @nexon-97 `camelCase` for local variables.

File: AppSample/Tracer.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+				throw new InvalidTraceException("Attempt to read trace result from tracer in invalid state!");
+			}
+
+			return result;
+		}
+	}
+
+	public MethodBase GetCurrentMethodInfo()
+	{
+		lock (StackFrameGetterLock)
```
Comment (RamanBut-Husaim) [inline AppSample/Tracer.cs]: @nexon-97 why are you using a lock there?

File: AppSample/TraceResult.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+		return true;
+	}
+
+	public BasicTreeNode<TraceComponent> GetThreadRootComponent(int ThreadId)
+	{
+		try
+		{
+			return ThreadsData[ThreadId].RootComponent;
+		}
+		catch (Exception)
```
Comment (RamanBut-Husaim) [inline AppSample/TraceResult.cs]: @nexon-97 Catch `Exception` is a bad idea. Why are you catching an exception right there?

File: AppSample/TraceResult.cs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	private ThreadData GetCurrentThreadData()
+	{
+		int ThreadId = Thread.CurrentThread.ManagedThreadId;
+		try
+		{
+			return ThreadsData[ThreadId];
+		}
+		catch (Exception)
```
Comment (RamanBut-Husaim) [inline AppSample/TraceResult.cs]: @nexon-97 The same stuff here.

File: AppSample/Tracer.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	public void StopTrace()
+	{
+		result.StopComponent();
+	}
+
+	public TraceResult GetTraceResult()
+	{
+		lock (ValidatorLock)
```
Comment (RamanBut-Husaim) [inline AppSample/Tracer.cs]: @nexon-97 What is the purpose of the lock right there?

**Comment by nexon-97** [comment]
Pushed modification based on your review, please recheck to confirm.

**Comment by RamanBut-Husaim** [comment]
@nexon-97 Waiting for the changes from the lesson.

**Comment by nexon-97** [comment]
Changes from the lesson have been pushed.

**Comment by RamanBut-Husaim** [comment]
@nexon-97 Accepted.

**Pull request merged**